chevah/python-package
# [#4837] cryptography updates.

#Scope
=====

Version 1.9 we are currently using on most OS'es is not known to have security issues, but it bundles OpenSSL 1.1.0f on Windows and OSX / MacOS, which does have a number of issues: ​https://www.cvedetails.com/vulnerability-list/vendor_id-217/product_id-383/version_id-223239/Openssl-Openssl-1.1.0f.html.

More so, newest pyOpenSSL requires newer `cryptography`. pyOpenSSL has fixed in latest versions bugs with a security impact. More at https://github.com/pyca/pyopenssl/blob/master/CHANGELOG.rst

Changes
=======

Updated `cryptography` for SLES 11 (and possibly RHEL 5) to latest version with support for OpenSSL 0.9.8, namely version **1.3.4**.

Updated `cryptography` to latest version **2.2.2** for all OS'es using 1.9.

Switched to using cffi-based modules for AIX, HP-UX and Solaris 11 with the following `cryptography` versions:
  * pristine **2.2.2** for AIX, with a minor patch for Python to find `ld_so_aix` & co.
  * slightly patched **2.2.2** for HP-UX (minor problems with CFLAGS and C syntax)
  * **2.2.2** / **1.8.2** / **1.6** for Solaris 11, depending on OpenSSL version and availability of EC crypto, more in https://github.com/chevah/python-package/pull/102/commits/6fc61a51dd32a18e22cf2f4a2e45d43295c06e70. A bit of pruning for compile flags is needed for 2.2.2.

**Drive-by fixes**:

  * updated `pip` and `setuptools` to latest stable versions: 9.0.3 and 39.0.1
  * updated `cffi` to latest version, 1.15.2, to avoid reinstalling when building `cryptography`'s and related issues on exotic platforms
  * happily hacked in support for OpenBSD 6.3 and LibreSSL 2.7.x
  * updated SQLite sources and Windows DLL to fix a nasty bug found upstream
  * updated `scandir` to latest version, as 1.5 PyPI archive was not handled well in the Unix world (CR+LF endings). `scandir` tests now pass on Debian Linux and Solaris 10 AMD64 too
  * call `ioscan` with full path on HP-UX to account for a $PATH with no `/usr/sbin`
  * updated `paver.conf` for Ubuntu 18.04
  * updated Solaris 11 installation on Jacu from 11.1 to 11.2 to match the Solaris 11 AMD64 installations. OpenSSL version was upgraded from 1.0.0j to 1.0.1h. 


How to try and test the changes
===============================

reviewers: @adiroiban 

Please review changes in `chevah_build` and `src/python`, everything else is related to updating upstream sources.
Run the automated tests.
Check out cffi-based modules in AIX / HP-UX / Solaris 11, especially `cryptography`.
Check failed `compat` tests on Solaris 11 SPARC.
